### ISSUE TITLE
Make the `qlever` script more portable

### ISSUE THREAD

**Issue by hannahbast** [ad-freiburg/qlever-control#16]
Continuing a discussion here started with @bradenmacdonald on https://github.com/ad-freiburg/qlever/pull/992 (the content of which was not directly related to this discussion, but just triggered it)

**Comment by hannahbast** [comment]
@bradenmacdonald Thank your for your thoughtful reply.

I also lean towards making an attempt at it with Python. We could start with a reduced version that provides just the basic actions `get-data`, `index`, `start`, `stop`, and the autocompletion. Then we could assess how that works out on various platforms and whether any principle problems arise.

I completely agree with you about the way the autocompletion is currently set up (not standard) and should be set up.

I am sceptical about the docker-based approach. Docker is great and should always be an option. But we don't want to force or almost force people to use it. Docker does come with a performance overhead, which is sometimes small but sometimes also not so small (depending on circumstances that are hard to understand).

Keeping the shell script but strictly sticking to POSIX standards would be an alternative. But I agree with you on the maintainability aspect. Shell scripts are fine when they are reasonably small and simple. The `qlever` script has already grown beyond that and will only grow more in the future.

An independent question is which format we should pick for the QLeverfiles. Right now, it shell-style variable assignment. Something like TOML would be more standard these days, but TOML doesn't really allow any sort of commands or does it?

Overall, we shouldn't shy away from a mahor revision. The `qlever` script in its current version was really just a proof of concept. And it proved very well that having a single `QLeverfile` is a good idea. The action-style approach (being able to write something like `qlever get-data index start`) also worked very well and conveniently. So the interface is nice, but there is no need to stick to how its currently implemented.

**Comment by bradenmacdonald** [comment]
> I also lean towards making an attempt at it with Python. We could start with a reduced version that provides just the basic actions get-data, index, start, stop, and the autocompletion. Then we could assess how that works out on various platforms and whether any principle problems arise.

I think this sounds like a good plan :)

> Keeping the shell script but strictly sticking to POSIX standards would be an alternative. But I agree with you on the maintainability aspect. Shell scripts are fine when they are reasonably small and simple. The qlever script has already grown beyond that and will only grow more in the future.

I completely agree. And just to be clear, I wasn't suggesting rewriting the current bash script to be more portable while keeping it as a shell script, but I was suggesting that we can still support using common commands like `wget` and shell features in the Qleverfiles themselves, since that seems to be convenient.

> An independent question is which format we should pick for the QLeverfiles. Right now, it shell-style variable assignment. Something like TOML would be more standard these days

If we go with python, I would suggest sticking to one of [the formats that Python's standard library can parse](https://docs.python.org/3/library/fileformats.html), so that users don't have to create a virtual environment or install dependencies in order to use the script. (That wouldn't be a big deal, but it's really nice if you can avoid it!) Of those, TOML seems like the best choice. I do like YAML as well, and it's perhaps more popular in general, but it requires a third party package to parse.

> but TOML doesn't really allow any sort of commands or does it?

I think you'd just have to specify things as plain strings in the toml file, and then the `qlever.py` control script can pass those strings to a shell, to achieve similar functionality to what you have today.

**Comment by hannahbast** [comment]
@bradenmacdonald There is now a first version of new script `qlever.py` based on Python. It already supports `get-data`, `index`, `start`, `stop`, and `status`, with and without Docker. There is a new `Qleverfile.ini` for the olympics dataset. I have also set up [a GitHub workflow with an end-to-end test](https://github.com/ad-freiburg/qlever-control/blob/python-qlever/.github/workflows/end-to-end-test.yml) on Ubuntu 22.04, which works fine.

The PR is #17 . Would you care to try this out on your Mac? You probably need to `pip3 install psutil`. You either need Docker or the compiled binaries. It would be great to try both (the YML of the end-to-end test linked above shows how the script is used, it's very similar to before).

The names `qlever.py` and `Qleverfile.ini` are preliminary. For now, I deliberately chose names that are the distinct from the previous one, so that I can work with both scripts in parallel.

Any questions, comments or suggestions you might have are welcome.

**Comment by hannahbast** [comment]
@bradenmacdonald There is now also an [end-to-end test for MacOS 12](https://github.com/ad-freiburg/qlever-control/blob/python-qlever/.github/workflows/end-to-end-test-macos.yml), which runs through fine.

**Comment by bradenmacdonald** [comment]
@hannahbast Very nice work - this is great to see! Thanks for letting me know :)

Of course, I am happy to help test it. I just did so, and left my findings on the PR.

Really glad to see the end to end test too 💯. Again, congrats - this looks like a very nice step forward!

**Comment by hannahbast** [comment]
@bradenmacdonald Thanks a lot for the feedback. I have incorporated your suggestions (better shebang, use curl instead of wget, more refined check if docker is installed -> stop should now work for you, check if binaries are installed). Eventually, the script should become as user-friendly as the old one (when it worked). The first step is to get it to work on a variety of platforms, including MacOS. When that is done, I will add more verbose help info, like in the old script.

**Concerning your binaries** While debugging the MacOS end-to-end test, I also encountered the error message you reported (libc++abi, U_FILE_ACCESS_ERROR, etc.). I don't know what exactly they mean, but I figured out that they occur when the binary is there, but libraries are missing or cannot be found. If you build the binaries as follows (in `qlever-code`, with a fresh clone of https://github.com/ad-freiburg/qlever), it should work:

```
brew install llvm@16
brew install conan@2
rm -rf build && mkdir build
(cd build && conan install .. -pr:b=../conanprofiles/clang-16-macos -pr:h=../conanprofiles/clang-16-macos -of=. --build=missing)
cmake -B $(pwd)/build -DCMAKE_TOOLCHAIN_FILE="$(pwd)/build/conan_toolchain.cmake" -DADDITIONAL_COMPILER_FLAGS="-Wall -Wextra -O3" -DCMAKE_BUILD_TYPE=Release -DUSE_PARALLEL=true -DRUN_EXPENSIVE_TESTS=false -DENABLE_EXPENSIVE_CHECKS=true -DCMAKE_CXX_COMPILER=clang++
source build/conanrun.sh
cmake --build $(pwd)/build --config Release -- -j 8
```

You can also copy the binaries from here: https://ad-research.cs.uni-freiburg.de/downloads/qlever/macos-12 . But even then you need the `conan install ...` and `source build/conanrun.sh` to run them, otherwise you get the error messages you reported.

**Docker** It's great to hear that `docker` works on your machine. My guess is that you use your Mac like a PC, with a screen. Our Mac is just a machine in a server room to which we can connect via `ssh`, so all we have is the command line. I tried installation via `docker` via `brew install --cask docker` as well as directly [from the DMG file](https://docs.docker.com/desktop/install/mac-install/). In both cases, the docker daemon is not running and I can't get it to run. That's also why the end-to-end test for macos-12 currently only tests the command with docker.USE_DOCKER=false. Any ideas?

**Comment by bradenmacdonald** [comment]
@hannahbast Awesome, thanks for incorporating those changes!

> they occur when the binary is there, but libraries are missing or cannot be found.

Oh right, thanks for the reminder. I'm pretty sure that's the issue - if one tries to use the binaries without first `source build/conanrun.sh`, then they will fail with that `U_FILE_ACCESS` error. @joka921 [told me that back in June](https://github.com/ad-freiburg/qlever/pull/989#issuecomment-1580286863) when I was testing the conan build, but I've mostly been using docker images since then so I forgot about it. That's basically the only thing I didn't like about the Conan setup was that you have to set up this environment before you can run the binary.

> Our Mac is just a machine in a server room to which we can connect via ssh, so all we have is the command line. ... Any ideas?

I don't have direct experience with this problem unfortunately. But I have heard good things about Colima and Orbstack, and I suspect one if not both of them would work for this case:
* [Colima](https://github.com/abiosoft/colima)
* [Orbstack post about headless use case](https://github.com/orbstack/orbstack/issues/194)

**Comment by hannahbast** [comment]
@bradenmacdonald Oh, wow, I did

```
brew install docker
brew install colima
colima start
```

And then everything just worked. I am truly amazed!

**Comment by bradenmacdonald** [comment]
Yay, I'm really glad it worked and was so simple!

**Comment by tfmorris** [comment]
What is the state of the Python script? I watched the Wikidata Modeling Days presentation and it looked super simple to just do `pip install qlever`, but that doesn't produce a working system for me. The installation runs without error, but when I try `qlever setup-config olympics` it looks like there is no `Qleverfiles` directory included with the installation.

Is this expected to work or is this still a work in progress? 

What is the best way to install today on an M3 Macbook Pro?

**Comment by hannahbast** [comment]
@tfmorris We are in the process of refactoring the script to make it more testable and so that it works on many different platforms. I will fix the pypi package. Just now, I am having trouble uploading the `Qleverfiles` directory (which as you correctly noticed, is currently missing).

What should work is: `git clone -b python-qlever https://github.com/qlever-control` in a directory of your choice and follow these simple instructions (essentially the same as in the talk we gave): https://github.com/ad-freiburg/qlever-control

**Comment by hannahbast** [comment]
@tfmorris I fixed the pypi package, please try `pip install qlever` again and let me know if it works.

**Comment by tfmorris** [comment]
@hannahbast That fixed the config file issue. Thanks for the quick turnaround!

There are a few issues with the templates themselves (e.g. RDF_FILES vs FILE_NAMES), but I'll file separate reports for those.

**Comment by hannahbast** [comment]
This is an old thread and it looks like all the issues here have been adressed in the meantime. If not, just open a new issue
